File: server/src/main/scala/kpn/api/common/data/Tagable.scala
package kpn.api.common.data

import kpn.api.custom.Tag
import kpn.api.custom.Tags

trait Tagable {

  def tags: Seq[Tag]

  def tagValue(key: String): Option[String] = {
    Tags.get(tags, key)
  }

  def hasTag(key: String, allowedValues: String*): Boolean = {
    Tags.has(tags, key, allowedValues: _*)
  }
}

File: backend/api/src/main/java/kpn/api/common/data/Tagable.java
package kpn.api.common.data;

import kpn.api.custom.Tag;
import kpn.api.custom.Tags;

import java.util.List;
import java.util.Optional;
import com.google.common.collect.ImmutableList;

public interface Tagable {

  List<Tag> tags();

  default Optional<String> tagValue(String key) {
    return Tags.get(tags(), key);
  }

  default ImmutableList<String> tagValues(String key) {
    return Tags.values(tags(), key);
  }

  default boolean hasTag(String key, String... allowedValues) {
    return Tags.has(tags(), key, allowedValues);
  }
}
